### PR TITLE
Reimplement entity equality

### LINKED ISSUE #15588: Query: Rewrite EntityEqualityRewrite

### PR DESCRIPTION
**Pull request by roji**
New implementation of entity equality rewriter, which follows exact entity (and anonymous) from roots to the comparison expression, for more precise rewriting.

Note that unlike the first attempt in #15823, the visitors in this PR are completely stateless.

Closes #15588

Additional scoped-out items:
* [ ] Rewrite Contains() (#15939)
* [ ] Owned entity support (#15936)
* [ ] GroupBy support (#15938)
* [ ] Cardinality-reducing methods accepting IQueryable<T> and returning superclass/subclass of T (https://github.com/aspnet/EntityFrameworkCore/pull/15920#discussion_r290529415)
* [ ] Possibly: complex EE cases (#15260)

### CODE REVIEW

**Comment by roji** [comment]
(will check failing tests tomorrow, in the meantime a review would be great)

**Comment by roji** [comment]
And one last comment before I'm off... Near the end of this refactor I realized that the EntityReferenceExpression nodes can probably be reduced during the first visit, obviating a second reduce-only pass. I'll look at that tomorrow.

**Comment by smitpatel** [comment]
Was gonna just that only. You don't need reducing visitor. Just top level method to unwrap the custom expression.

**Comment by roji** [comment]
@smitpatel yep :) It took me a while but I did get it in the end...

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs [hunk 1]
```diff
@@ -1749,7 +1749,7 @@ public virtual Task OrderBy_Skip_Last_gives_correct_result(bool isAsync)
                 entryCount: 1);
         }
 
-        [ConditionalFact(Skip = "Issue#15588")]
+        [ConditionalFact(Skip = "Null TypeMapping in Sql Tree")]
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs]: EE supposed to convert Queryable<TEntity>.Contains(TEntity) to Contains over PK when single PK column

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.cs [hunk 2]
```diff
@@ -429,7 +443,17 @@ public virtual Task Entity_equality_null(bool isAsync)
...
-        [ConditionalTheory(Skip = "Issue#15588")]
+        [ConditionalTheory]
+        [MemberData(nameof(IsAsyncData))]
+        public virtual Task Entity_equality_null_composite_key(bool isAsync)
+            => AssertQuery<OrderDetail>(
+                isAsync,
+                odt =>
+                    from od in odt
+                    where od == null
+                    select (object)od.ProductID);
```
Comment (smitpatel) [inline test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.cs]: Is object cast related to the query? If not, remove

File: src/EFCore/Query/Pipeline/QueryOptimizingExpressionVisitor.cs [hunk 3]
```diff
@@ -22,6 +22,7 @@ public Expression Visit(Expression query)
             query = new AllAnyToContainsRewritingExpressionVisitor().Visit(query);
             query = new GroupJoinFlatteningExpressionVisitor().Visit(query);
             query = new NullCheckRemovingExpressionVisitor().Visit(query);
+            query = new EntityEqualityRewritingExpressionVisitor2(_queryCompilationContext).Visit(query);
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/QueryOptimizingExpressionVisitor.cs]: drop 2

File: src/EFCore/Query/Pipeline/QueryableMethodTranslatingExpressionVisitor.cs [hunk 4]
```diff
@@ -430,7 +430,7 @@ protected override Expression VisitMethodCall(MethodCallExpression methodCallExp
                     }
                 }
 
-                throw new NotImplementedException();
+                throw new NotImplementedException("Unhandled method: " + methodCallExpression.Method.Name);
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/QueryableMethodTranslatingExpressionVisitor.cs]: Nice!

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+            }
+
+            protected override Expression VisitConstant(ConstantExpression constantExpression)
+                => constantExpression.IsEntityQueryable()
+                    ? new EntityReferenceExpression(constantExpression, Model.FindEntityType(((IQueryable)constantExpression.Value).ElementType))
+                    : (Expression)constantExpression;
+
+            protected override Expression VisitNew(NewExpression newExpression)
+            {
+                var newNew = (NewExpression)base.VisitNew(newExpression);
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: use visitedExpression.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 6]
```diff
@@ -0,0 +1,636 @@
...
+
+                return (newExpression.Members?.Count ?? 0) == 0
+                    ? (Expression)newNew
+                    : new EntityReferenceExpression(newNew, newNew.Members
+                        .Select((m, i) => (Member: m, Index: i))
+                        .ToDictionary(
+                            mi => mi.Member.Name,
+                            mi => Visit(newExpression.Arguments[mi.Index])));
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Visiting arguments twice

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+            }
+
+            protected override Expression VisitMember(MemberExpression memberExpression)
+            {
+                var newMemberExpression = (MemberExpression)base.VisitMember(memberExpression);
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Visit inner and reconstruct

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+            protected override Expression VisitExtension(Expression expression)
+                => expression is EntityReferenceExpression ? expression : base.VisitExtension(expression);
+
+            /// <summary>
+            ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+            ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+            ///     any release. You should only use it directly in your code with extreme caution and knowing that
+            ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+            /// </summary>
+            protected static Expression CreateKeyAccessExpression(
```
Comment (maumar) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: DRY (very similar code is in NavigationExpansionHelpers). Can be done in future checkin but probably good idea to add TODO, so we dont forget ;)

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+                if (arguments.Count == 0)
+                {
+                    return methodCallExpression.Update(Visit(methodCallExpression.Object), Array.Empty<Expression>());
+                }
+
+                // Methods with one lambda argument that has one parameter, and with a typed first argument (source) are rewritten automatically
+                // (e.g. Where(), FromSql(), Average()
+                var newArguments = new Expression[arguments.Count];
+                var lambdaParamCount = arguments.Count(a => GetLambdaOrNull(a) != null);
+                var singleLambda = arguments.Select(GetLambdaOrNull).SingleOrDefault(l => l != null);
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Is SingleOrDefault safe?

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 10]
```diff
@@ -0,0 +1,636 @@
...
+                newSource = newArguments[0] = Visit(arguments[0]);
+                if ((lambdaParamCount == 0 || singleLambda.Parameters.Count == 1)
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Isn't LambdaParamCount == 0 equivalent to singleLambda being null

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 11]
```diff
@@ -0,0 +1,636 @@
...
+
+            /// <summary>
+            /// The underlying expression being wrapped.
+            /// </summary>
+            [NotNull]
+            public Expression Underlying { get; }
+
+            public override Type Type => Underlying.Type;
+
+            public override bool CanReduce => true;
```
Comment (maumar) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: i would not make it reducible, but rather prune the nodes explicitly with a dedicated visitor (as you are already doing)

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 12]
```diff
@@ -0,0 +1,636 @@
...
+                // Unknown method - still need to visit all arguments
+                for (var i = 1; i < arguments.Count; i++)
+                {
+                    newArguments[i] = Visit(arguments[i]);
+                }
+
+                return methodCallExpression.Update(Visit(methodCallExpression.Object), newArguments);
+            }
+
+            protected static bool IsQueryableOrEnumerable(Type type, out Type elementType)
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Look for TryGetElementType

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 13]
```diff
@@ -0,0 +1,636 @@
...
+            protected virtual Expression VisitJoinMethodCall(MethodCallExpression methodCallExpression)
+            {
+                var arguments = methodCallExpression.Arguments;
+
+                if (arguments.Count != 5)
+                {
+                    return base.VisitMethodCall(methodCallExpression);
+                }
+
+                var (newOuter, newInner, outerKeySelector, innerKeySelector, resultSelector) = (
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Break down into individual lines. readability is reduced when there are too many of assignments

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 12 above)
Comment (maumar) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: or better yet, TryGetSequenceType

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 14]
```diff
@@ -0,0 +1,636 @@
...
+            public override Expression Reduce() => Underlying;
+
+            [CanBeNull]
+            public IEntityType EntityType { get; }
+
+            [CanBeNull]
+            public INavigation LastNavigation => EntityType == null ? null : _lastNavigation;
+
+            [CanBeNull]
+            private readonly INavigation _lastNavigation;
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Should store navigation chain instead. For Aggregate we need to trim down navigation chain upto the root

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 15]
```diff
@@ -0,0 +1,636 @@
...
+
+            [CanBeNull]
+            public Dictionary<string, Expression> AnonymousType { get; }
+
+            public bool SubqueryTraversed { get; }
+
+            public bool IsAnonymousType => AnonymousType != null;
+            public bool IsEntityType => EntityType != null;
+
+            public EntityReferenceExpression(Expression underlying, Dictionary<string, Expression> anonymousType)
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Should store using Member rather than string perhaps

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 16]
```diff
@@ -0,0 +1,636 @@
...
+                       ? entityWrapper.TraverseProperty(newMemberExpression.Member.Name, newMemberExpression)
+                       : newMemberExpression;
+            }
+
+            protected override Expression VisitBinary(BinaryExpression binaryExpression)
+            {
+                if (binaryExpression.NodeType == ExpressionType.Equal || binaryExpression.NodeType == ExpressionType.NotEqual)
+                {
+                    var (newLeft, newRight) = (Visit(binaryExpression.Left), Visit(binaryExpression.Right));
+                    return RewriteEquality(binaryExpression.NodeType == ExpressionType.Equal, newLeft, newRight)
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: It would be more beneficial if you visited inside the rewriting function.
Function would visit left & right & find wrappers. If they match (with or without removing matching navigations which can be removed using unwrapnav), you can rewrite key equality on original nodes directly.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 12 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: but we should convert everything Queryable in advance too.

Comment (maumar) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: #15931

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.cs (hunk 2 above)
Comment (roji) [reply test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.cs]: Does not compile otherwise, because "Cannot implicitly convert type 'System.Linq.IQueryable<int>' to 'System.Linq.IQueryable<object>'.". Probably because `ObjectDetail.ProductID` is an int (where Customer has a string ID?).

Changing to simply return OrderDetails itself instead of its ID, seems OK.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 5 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Changed to `visitedNewExpression` - although elsewhere the visited are usually referred to as newLambda, newMethodCall, etc. Not a big deal.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 6 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Good catch, thanks!

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 7 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: OK, but any reason for this? Anything wrong with the way it was done before?

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 8 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: I actually moved this code from ExpressionExtensions.cs since it wasn't being used by anyone. Should I revert, do you need this exactly?

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 9 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Oops, was somehow under the impression SingleOrDefault returns null for multiple... Correcting.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 10 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: The point was to detect either no lambdas, or a single lambda with a single parameter. But as you noted above SingleOrDefault throws on multiple lambda parameters, so I changed the logic

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 12 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Thanks, cool that these exist! Although we should rename to GetSequenceType (and maybe add bool-returning TryGetSequenceType) to align with C# standards.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 13 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Agree on second look...

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 14 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: You mean for [`Queryable.Aggregate`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.queryable.aggregate?view=netframework-4.8#System_Linq_Queryable_Aggregate__3_System_Linq_IQueryable___0____1_System_Linq_Expressions_Expression_System_Func___1___0___1___System_Linq_Expressions_Expression_System_Func___1___2___)?

I don't see why EE specifically (not other components) would need the entire navigation chain... Doesn't the entity type simply flow from the source into the accumulator function, and the accumulator flows from the seed to the result selector, which projects to the result of the entire function?

Either way, do we see ourselves translating Aggregate anytime soon? Even if so, maybe we should defer it until then?

Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Are you referring to GroupBy specifically, and how to handle the resulting `IEnumerable<IGrouping<TKey,TSource>>`? I was thinking of handling this in a similar way as anonymous objects (but haven't thought it through yet).

Either way, can we defer this until GroupBy is actually working? Could also be post 3.0.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 15 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: It's a nice idea but things get a little bit complicated, as strings still need to be provided from EF.Property accesses... We could have two `TraverseProperty` overloads - one with string and one with MemberInfo - but it seems it doesn't make the code better. If you feel strongly about it I can do it.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 16 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: This is actually how the older (stateful) version worked. But the rewriting can abort, e.g. because neither side has known entities (or one side is an anonymous object) - at that point we've already visited, and to avoid double visitation we have to pass the visited expressions back out, which is why there used to be two `ref` parameters. You didn't like it :)

Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: We don't have to worry about how older version worked. It depends on the logic or requirement of any rewrite.
If you can look at visited expression and rewrite something from them only, then your helper method either returns rewritten expression or null at which you can create default expression.
But if your helper method requires access to old or requires visiting then helper method can deal with those expressions fully and always return a non-null result (optimized or default). At the end of the day, it's just a private (or protected) method so it is not black and white.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 15 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Anonymous types must have members and would be accessed via members only hence suggestion. But I see why we need strings (though they will not be used for anonymous type member access).
You can leave this as is.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 14 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Owned entities aggregate.
`c1.OwnedNav1.OwnedNav2 == c2.OwnedNav1.OwnedNav2` should be equivalent to `c1.Id == c2.Id`

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 5 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: we use expressionType ending with expression as parameter name (e.g. memberExpression)
While visited one can be suffixed with either new or visited or updated (I have less preference for new) but it still captures the expressiontype name. Certain expression type names are fine without expression ending but not all. `newMethodCall` would not be confusing but newMember is confusing whether it is a memberInfo or memberExpression. In IDE name shouldn't matter much since auto-complete will give you suggestions but reading on github becomes harder. So I tend to more verbose if it makes easier. (Haven't you seen @maumar's 50 character long variable names?) For consistency in ExpressionVisitors, follow the guidelines in first 2 lines.

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.cs]: Do you need to use `AssertQueryScalar` instead?
AssertQuery works for reference types only.

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs (hunk 1 above)
Comment (smitpatel) [reply test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs]: You can scope this out of this PR and file an issue and link it here.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 7 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: I did not realize we are passing updatedExpression in first branch.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs [hunk 17]
```diff
@@ -0,0 +1,636 @@
...
+            [CanBeNull]
+            private readonly INavigation _lastNavigation;
+
+            [CanBeNull]
+            public Dictionary<string, Expression> AnonymousType { get; }
+
+            public bool SubqueryTraversed { get; }
+
+            public bool IsAnonymousType => AnonymousType != null;
+            public bool IsEntityType => EntityType != null;
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: It is interesting that EntityType can be null too since entity reference should be referencing an entityType.
Probably, we are re-using same expression to store the shape & reference itself rather than composing.
I have some idea in mind. But let's discuss it after this PR is merged.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 16 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: I don't understand...

You're proposing that I visit the operands in VisitBinary instead of in RewriteEquality. If RewriteEquality cannot rewrite (e.g. because the operands aren't entity-referencing), we still need to construct a new BinaryExpression from them, right? Since this needs to happen in VisitBinary and not in RewriteEqualiy, that would mean we need to pass the visited operands out from RewriteEquality back to VisitBinary.

This can be done by having RewriteEquality use ref parameters, but I don't really see the advantage over the current factoring...

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 17 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Yeah, the name `EntityReferenceExpression` is imprecise here as it can refer to an anonymous object type. It was previously called `TypeInfoWrapperExpression` or similar. We can rename if this bothers you.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 14 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Ah, I see now - make sense. That's special handling for owned entities (i.e. doesn't occur in any other case), right?

This shouldn't be too hard to do, but maybe it's better to wait until owned entities are implemented (#15285)? Or should they already be working?

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 16 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: > Since this needs to happen in VisitBinary and not in RewriteEqualiy

That is not true.
If one finds `a == b` then just pass it off to RewriteEquality function. The function will visit children, rewrite if needed. If cannot rewrite then construct the updated binary expression and return.

Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: But the point of RewriteEquality is to contain the common logic between VisitBinary and VisitMethodCall (for `object.Equals()`), but keep the non-common logic in the latter. You think it's preferable for RewriteEquality to know binary equality vs. the two types of `object.Equals()`? 

BTW the same would have to be true of unpacking the operands - there are 3 options. These seem like they really belong in VisitBinary/VisitMethodCall (otherwise they'd have to be duplicated there anyway, to recognize `object.Equals()`).

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 17 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: You can leave name as is. If we split it out in 2 then name won't be issue, else we can decide during API review.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 14 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: We can defer.

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 17 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: OK.

BTW are we treating this as a public surface API?? I was sure this would be pubternal...

Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: API review is for that only. 😉

Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: That's why I asked... :) Seems a bit extreme to expose so much gritty internals as public!

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 14 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: #15285

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 8 above)
Comment (maumar) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: its fine, when someone needs a similar code they can do the DRYing

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 11 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: This seems to be necessary for ReplacingExpressionVisitor - if we have nested lambdas, then parameter expressions wrapped by a previous replace for an outer lambda are encountered during the replace for an inner lambda.

By the way, I think ReplacingExpressionVisitor may need to have the following, to ensure that it doesn't make extension nodes disappear:

```c#
protected override Expression VisitExtension(Expression expression)
{
    // Leave extension nodes in place unless a replace actually occurs within their reduced form.
    if (!expression.CanReduce)
    {
        return base.VisitExtension(expression);  // This throws
    }

    var reduced = expression.Reduce();
    var visited = Visit(reduced);

    return reduced == visited ? expression : visited;
}
```

Here's a test the demonstrates the issue:

```c#
[ConditionalTheory]
[MemberData(nameof(IsAsyncData))]
public void Inner_parameter_in_nested_lambdas_gets_preserved(bool isAsync)
{
    AssertQuery<Customer>(
        isAsync,
        cs => cs.Where(c => c.Orders.Where(o => c == new Customer { CustomerID = o.CustomerID }).Count() > 0),
        entryCount: 90);
}
```

Basically, as explained above the `c` inside the nested where is first rewritten to an EntityReferenceExpression when we reach the outer lambda. But when we reach the inner lambda, the 2nd ReplacingExpressionVisitor sees the extension node wrapping `c` and reduces it - making it disappear.

I tried to make this change but it seems to break lots of other tests for various unrelated reasons.

@smitpatel does this make sense? Should I open an issue to track?

File: test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs (hunk 1 above)
Comment (roji) [reply test/EFCore.Specification.Tests/Query/SimpleQueryTestBase.ResultOperators.cs]: #15939

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs (hunk 11 above)
Comment (smitpatel) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Override VisitChildren method in your custom expression. ReplacingExpressionVisitor is working correctly.

Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor2.cs]: Oh, I somehow missed that entirely, thanks!

**Comment by maumar** [comment]
@roji you can also remove entity equality code from nav expansion (NavigationExpandingVisitor.VisitBinary + CreateNullComparisonArguments) to make sure the new logic covers all the scenarios

**Comment by roji** [comment]
@maumar done, 0 test failures after removing that code :)

**Comment by roji** [comment]
OK guys, I think it's done.

The visitor now reduces nodes as it does its pass, no more 2nd reducing pass. There are also a ton more fixes and cleanups - it may need to be reviewed fully again. Sorry for all the back-and-forths on this, hopefully my future query PRs will be more quick and direct to the point.

File: src/EFCore/Properties/CoreStrings.resx [hunk 18]
```diff
@@ -1168,4 +1168,7 @@
   <data name="NoNavigation" xml:space="preserve">
     <value>There is no navigation on entity type '{entityType}' associated with the foreign key {foreignKey}.</value>
   </data>
-</root>
\ No newline at end of file
+  <data name="SubqueryWithCompositeKeyNotSupported" xml:space="preserve">
+    <value>This query would cause multiple evaluation of a subquery because entity '{entityType}' has a composite key. Rewrite your query avoiding the subquery.</value>
```
Comment (roji) [inline src/EFCore/Properties/CoreStrings.resx]: @divega this scenario - entity equality on entities with composite keys over a subquery - has never been translated as it would cause multiple evaluation of the subquery. As we no longer have client evaluation, I'm proposing to throw an informative NotSupportedException that would guide the user on avoiding the pattern, does this make sense?

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor.cs [hunk 19]
```diff
@@ -0,0 +1,739 @@
...
+                        && (returnElementType == sourceElementType || sourceElementType == null))
+                    {
+                        return newSourceWrapper.Update(
+                            methodCallExpression.Update(null, newArguments));
+                    }
+
+                    // If the source type is an IQueryable over the return type, this is a cardinality-reducing method (e.g. First).
+                    // These don't flow the last navigation. In addition, these will be translated into a subquery, and we should not
+                    // perform entity equality rewriting if the entity type has a composite key.
+                    if (methodCallExpression.Method.ReturnType == sourceElementType)
```
Comment (maumar) [inline src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor.cs]: what about: ctx.Customers.OfType<VipCustomer>().FirstOrDefault<Customer>(), i.e. when type argument of cardinality reducing operator is explicitly defined and different than one on source?

Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor.cs]: Hmm, interesting - I've never seen someone call FirstOrDefault with an explicit type parameter, any specific value in doing so?

Comment (maumar) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor.cs]: can't think of anything reasonable where the result would be an entity but I'm working on a somewhat similar problem in different area (#15535), where the First method was typed as object. Might be worth adding test just in case. We have some inheritance in GearsOfWar model (Faction entity) that can be used for that purpose.

File: src/EFCore/Properties/CoreStrings.resx (hunk 18 above)
Comment (divega) [reply src/EFCore/Properties/CoreStrings.resx]: Yes, sure. Though not sure I understand the scenario. Is this something that in the future we could translate with EXISTS?

Comment (roji) [reply src/EFCore/Properties/CoreStrings.resx]: Opened #15972 to track, including a full explanation - we should be able to translate this (not with EXISTS though). Note that this wasn't supported before (so it would be an enhancement rather than a regression fix).

File: src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor.cs (hunk 19 above)
Comment (roji) [reply src/EFCore/Query/Pipeline/EntityEqualityRewritingExpressionVisitor.cs]: Am deferring this for now (made a note at the top post) so that I can merge this PR quickly. Will get around to it soon.
